### ISSUE TITLE
How could we use GPU to train paccmann?

### ISSUE THREAD

**Issue by DBinary** [PaccMann/paccmann_predictor#25]
When we used the default parameter as you provided , We found that the mode was CPU to train our model . 
So how could we use GPU to train paccmann with our own data? Which parameter should we revise？

**Comment by jannisborn** [comment]
HI, 
It should be displayed in the logfiles. Search for `device` in the training script here: https://github.com/PaccMann/paccmann_predictor/blob/e1c36bc06778d06f95b9a328e2be6306b05e761e/examples/IC50/train_paccmann.py#L144

The device for the dataset you can set it via `dataset_device` in the config json.

The model will be on GPU if and only if a GPU is available. See here: https://github.com/PaccMann/paccmann_predictor/blob/e1c36bc06778d06f95b9a328e2be6306b05e761e/examples/IC50/train_paccmann.py#L187

If the model trained on CPU, it most likely means that pytorch did not detect a gpu on your machine.

**Comment by DBinary** [comment]
Thank you. Perhaps I was mistaken.

But when I wanted to use our own data to train a new paccmann model, there were some more bugs here:
`
INFO:train_air_0:Start data preprocessing...
ERROR:pytoda.datasets.smiles_dataset:Since you provided a smiles_language, the following parameters to this class will be ignored: canonical, augment, kekulize, all_bonds_explicit, selfies, sanitize, all_hs_explicit, remove_bonddir, remove_chirality, randomize, add_start_and_stop, padding, padding_length, device.
Here are the problems:
ERROR:pytoda.datasets.smiles_dataset:Provided arg padding:True does not match the smiles_language value: False NOTE: smiles_language value takes preference!!
ERROR:pytoda.datasets.smiles_dataset:Provided arg device:cpu does not match the smiles_language value: cuda NOTE: smiles_language value takes preference!!
ERROR:pytoda.datasets.smiles_dataset:To get rid of this, adapt the smiles_language *offline*, feed itready for intended usage, and adapt the constructor args to be identical with their equivalents in the language object
ERROR:pytoda.datasets.smiles_dataset:Since you provided a smiles_language, the following parameters to this class will be ignored: canonical, augment, kekulize, all_bonds_explicit, selfies, sanitize, all_hs_explicit, remove_bonddir, remove_chirality, randomize, add_start_and_stop, padding, padding_length, device.
Here are the problems:
ERROR:pytoda.datasets.smiles_dataset:Provided arg padding:True does not match the smiles_language value: False NOTE: smiles_language value takes preference!!
ERROR:pytoda.datasets.smiles_dataset:Provided arg device:cpu does not match the smiles_language value: cuda NOTE: smiles_language value takes preference!!
ERROR:pytoda.datasets.smiles_dataset:To get rid of this, adapt the smiles_language *offline*, feed itready for intended usage, and adapt the constructor args to be identical with their equivalents in the language object
/home/DAIR/hul/DrugAI/paccmann_predictor/examples/IC50/train_paccmann.py:178: DtypeWarning: Columns (1) have mixed types.Specify dtype option on import or set low_memory=False.
  iterate_dataset=False,
`
What should I do to fix this bug? Regenerating a new smile language obejct?

**Comment by DBinary** [comment]
Thank you. Perhaps I was mistaken.

But when I wanted to use our own data to train a new paccmann model, there were some more bugs here:

``
(paccmann_predictor) hul@air-debug-1:~/DrugAI/paccmann_predictor$ sh train_air_model.sh
INFO:train_air_0:Start data preprocessing...
ERROR:pytoda.datasets.smiles_dataset:Since you provided a smiles_language, the following parameters to this class will be ignored: canonical, augment, kekulize, all_bonds_explicit, selfies, sanitize, all_hs_explicit, remove_bonddir, remove_chirality, randomize, add_start_and_stop, padding, padding_length, device.
Here are the problems:
ERROR:pytoda.datasets.smiles_dataset:Provided arg padding:True does not match the smiles_language value: False NOTE: smiles_language value takes preference!!
ERROR:pytoda.datasets.smiles_dataset:Provided arg device:cpu does not match the smiles_language value: cuda NOTE: smiles_language value takes preference!!
ERROR:pytoda.datasets.smiles_dataset:To get rid of this, adapt the smiles_language *offline*, feed itready for intended usage, and adapt the constructor args to be identical with their equivalents in the language object
ERROR:pytoda.datasets.smiles_dataset:Since you provided a smiles_language, the following parameters to this class will be ignored: canonical, augment, kekulize, all_bonds_explicit, selfies, sanitize, all_hs_explicit, remove_bonddir, remove_chirality, randomize, add_start_and_stop, padding, padding_length, device.
Here are the problems:
ERROR:pytoda.datasets.smiles_dataset:Provided arg padding:True does not match the smiles_language value: False NOTE: smiles_language value takes preference!!
ERROR:pytoda.datasets.smiles_dataset:Provided arg device:cpu does not match the smiles_language value: cuda NOTE: smiles_language value takes preference!!
ERROR:pytoda.datasets.smiles_dataset:To get rid of this, adapt the smiles_language *offline*, feed itready for intended usage, and adapt the constructor args to be identical with their equivalents in the language object
/home/DAIR/hul/DrugAI/paccmann_predictor/examples/IC50/train_paccmann.py:178: DtypeWarning: Columns (1) have mixed types.Specify dtype option on import or set low_memory=False.
  iterate_dataset=False,
INFO:train_air_0:Training dataset has 46748 samples, test set has 46748.
``

What should I do to fix this bug? Regenerating a new smile language obejct?

**Comment by DBinary** [comment]
Thank you. Perhaps I was mistaken.

But when I wanted to use our own data to train a new paccmann model, there were some more bugs here:

INFO:train_air_0:Start data preprocessing...
ERROR:pytoda.datasets.smiles_dataset:Since you provided a smiles_language, the following parameters to this class will be ignored: canonical, augment, kekulize, all_bonds_explicit, selfies, sanitize, all_hs_explicit, remove_bonddir, remove_chirality, randomize, add_start_and_stop, padding, padding_length, device.
Here are the problems:
ERROR:pytoda.datasets.smiles_dataset:Provided arg padding:True does not match the smiles_language value: False NOTE: smiles_language value takes preference!!
ERROR:pytoda.datasets.smiles_dataset:Provided arg device:cpu does not match the smiles_language value: cuda NOTE: smiles_language value takes preference!!
ERROR:pytoda.datasets.smiles_dataset:To get rid of this, adapt the smiles_language *offline*, feed itready for intended usage, and adapt the constructor args to be identical with their equivalents in the language object
ERROR:pytoda.datasets.smiles_dataset:Since you provided a smiles_language, the following parameters to this class will be ignored: canonical, augment, kekulize, all_bonds_explicit, selfies, sanitize, all_hs_explicit, remove_bonddir, remove_chirality, randomize, add_start_and_stop, padding, padding_length, device.
Here are the problems:
ERROR:pytoda.datasets.smiles_dataset:Provided arg padding:True does not match the smiles_language value: False NOTE: smiles_language value takes preference!!
ERROR:pytoda.datasets.smiles_dataset:Provided arg device:cpu does not match the smiles_language value: cuda NOTE: smiles_language value takes preference!!
ERROR:pytoda.datasets.smiles_dataset:To get rid of this, adapt the smiles_language *offline*, feed itready for intended usage, and adapt the constructor args to be identical with their equivalents in the language object
/home/DAIR/hul/DrugAI/paccmann_predictor/examples/IC50/train_paccmann.py:178: DtypeWarning: Columns (1) have mixed types.Specify dtype option on import or set low_memory=False.
  iterate_dataset=False,


What should I do to fix them? Regenerating a new smile language obejct?

**Comment by jannisborn** [comment]
1. It's not an error strictly speaking, maybe we should change `ERROR:` to `WARNING:`... If the log message reflects what you intended to do, you can safely proceed.
2. If not, you can fix this as suggested in the last sentence: `ERROR:pytoda.datasets.smiles_dataset:To get rid of this, adapt the smiles_language offline, feed itready for intended usage, and adapt the constructor args to be identical with their equivalents in the language object.`


The reason here is that there is a bunch of SMILES related hyperparameters (padding, device, canonical, augment etc) that can be set *either* in the SMILES Language *or* in the dataset. The training script passes the argument from your config to the dataset constructor. If you dont provide a smiles language, everything goes fine, but if you provide one, then this language also contains specifications of these arguments that are passed to the dataset. If they dont match, a warning is raised which is what you see here. 

Hope this helps

**Comment by DBinary** [comment]
Thank you so much,you solved all questions I had.

**Comment by jannisborn** [comment]
Great, closing this now. Good luck with the project!
